tutorials/basic-topics/sorting/merge-sort.md

```diff
@@ -11,4 +11,519 @@ keywords:
+
+
+![image](https://cdn.kastatic.org/ka-perseus-images/5fcbebf66560d8fc490de2a0d8a0e5b1d65c5c54.png)
+
+
+
+### **Mergesort is also useful for sorting linked lists**
+
+## [**Leetcode Problem 148_Sort List**](https://leetcode.com/problems/sort-list/)
+
+**Problem Statement** 
+
+`Given the head of a linked list, return the list after sorting it in ascending order.`
+
+**Test Cases**
+ 
+ *Test Case 1*
+
+```
+Input: head = [4,2,1,3]
+Output: [1,2,3,4]
+```
+![image](https://assets.leetcode.com/uploads/2020/09/14/sort_list_1.jpg)
+ 
+*Test Case 2*
+
+```
+Input: head = [-1,5,3,4,0]
+Output: [-1,0,3,4,5]
+```
+
+Here we can follow both top down or bottom up merge sort . I have already discussed the algorithms in detailed way for array .
+
+### **Intuition for the problem**
+
+```
+3->8->12->null
+1->10->23->null
+
+After merging we get :
+
+1->3->8->10->12->23->null.
+
+Of course the two lists must are ordered respectively.So we can only start to merge two lists that only 
+have one element,then we get a ordered list that have two element,do this again (that is "recursion").
+
+    Bottom Up Approach 
+
+    -1-> 5->  3-> 4->   0-> null
+    \  /       \ /         /
+   -1->5->null 0->3->4->null 
+      \        /         
+-1-> 0->3-> 4-> 5->null   (sorted) 
+
+    Top Down Approach
+
+    -1-> 5->  3-> 4-> 0-> null
+         /        \ 
+-1-> 5->null      3-> 4->0->null
+    /  \          /      \  
+-1->null 5->null 3->null  4->0->null
+     \   /        \        /     \
+    -1-> 5->null  3->null 4->null 0->null
+      \             \        \   /
+                 3->null    0-> 4->null
+        \              \      /  
+                    0->3->4->null
+          \          /
+        -1->0->3->4->5->null    (sorted)
+
+```
+
+
+**Leetcode Implementation of Linked List in cpp**
+```
+/*
+* Definition for singly-linked list.
+* struct ListNode {
+*     int val;
+*     ListNode *next;
+*     ListNode() : val(0), next(nullptr) {}
+*     ListNode(int x) : val(x), next(nullptr) {}
+*     ListNode(int x, ListNode *next) : val(x), next(next) {}
+* };
+*/
+
+ ```
+
+
+## **Approach for top down Merge sort of Linked list**
+
+
+*1. Keep recursively dividing the list until there is only one node in the linked list.* 
+
+*2. Sort each sublist and merge each sorted sublist in a new array.*
+
+
+### **Steps to apply merge sort to a Linked list** 
+    
+**Note. Implementation is basically based on Merge Sort on an array as discussed above.** 
+
+1. Divide the linked list into two equal parts until when only one element is left
+2. To Divide, we need to  find mid in linked ling using [tortoise and hare](https://dev.to/alisabaj/floyd-s-tortoise-and-hare-algorithm-finding-a-cycle-in-a-linked-list-39af) method 
+3. Finally, merge the left and the right nodes of linked list
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+class Solution {
+public:
+    void mergeSort(ListNode **head){
```

Comment:
Format the code in IDE and add it. 
Remove unnecessary comments.

Reply:
ohk